### PR TITLE
[alea] Handle the case of zero/one measurements reponsibly.

### LINKED ISSUE #462: [ALEA] Inf/NaN in autocorr_acc result
Division by zero occurs when `autocorr_acc` tries to normalize variance at the uppermost level.

https://github.com/ALPSCore/ALPSCore/pull/446#issuecomment-365810182

```c++
#include <iostream>
#include <vector>

#include <alps/alea/autocorr.hpp>
#include "dataset.hpp"

int main()
{
    alps::alea::autocorr_acc<double> acc(2);
    for (size_t i = 0; i != twogauss_count; ++i)
        acc << std::vector<double>{twogauss_data[i][0], twogauss_data[i][1]};

    alps::alea::autocorr_result<double> res = acc.result();

    std::cout << "nlevel = " << res.nlevel() << std::endl;
    std::cout << "level(9) = " << res.level(9) << std::endl;

    return 0;
}
```
```
nlevel = 10
level(9) =   1.00718
-0.497835 +-  inf
-nan
```

### PR DESCRIPTION
**Pull request by mwallerb**
Fixes #462

### CODE REVIEW

File: alea/include/alps/alea/complex_op.hpp [hunk 1]
```diff
@@ -201,6 +201,29 @@ class complex_op
                           inv_t * x.imre(), inv_t * (x.imim() + s));
     }
 
+    friend bool isnan(complex_op x)
```
Comment (egull) [inline alea/include/alps/alea/complex_op.hpp]: please comment your functions

File: alea/include/alps/alea/complex_op.hpp [hunk 2]
```diff
@@ -201,6 +201,29 @@ class complex_op
...
+    }
+
+    friend bool isinf(complex_op x)
+    {
+        // An object cannot be inf and nan at the same time, for multi-
+        // component objects nan takes precedence
+        if (isnan(x))
+            return false;
+
+        return std::any_of(&x.vals_[0][0], &x.vals_[2][0],
```
Comment (egull) [inline alea/include/alps/alea/complex_op.hpp]: are you sure about the 2? Should it be a 1? (same everywhere else in this block)

Comment (mwallerb) [reply alea/include/alps/alea/complex_op.hpp]: No it's 2 here (ranges are exclusive).

File: alea/src/variance.cpp [hunk 3]
```diff
@@ -27,17 +27,38 @@ void var_data<T,Str>::reset()
 template <typename T, typename Str>
 void var_data<T,Str>::convert_to_mean()
 {
+    // This also works for count_ == 0
```
Comment (krivenko) [inline alea/src/variance.cpp]: Hmm. What's gonna happen if `count_ == 0`?
`data_`, `data2_` and `nunbiased` will all be nan, no?

Comment (mwallerb) [reply alea/src/variance.cpp]: Exactly: `data_` and `data2_`, that is, the mean and variance will be `nan`.  This is correct conceptually, since both are `0./0.`.

Comment (krivenko) [reply alea/src/variance.cpp]: Fine. I agree.
Could you please amend `operator==` to return `true` immediately if `(r1.count() == 0) && (r2.count() == 0)`?

Comment (mwallerb) [reply alea/src/variance.cpp]: Oh yes, good point.

Comment (galexv) [reply alea/src/variance.cpp]: Along the same lines, and also as Emanuel(@egull) mentioned: there are no tests for `operator==`: how do we know that it works? ;)  And we have other tests that depend on it working.
(But possibly it should be a separate issue deserving a separate PR)

**Comment by krivenko** [comment]
@mwallerb Just in case, it seems like we are not directly affected by this issue at the moment. No rush, therefore.

**Comment by krivenko** [comment]
@mwallerb I tried to rebuild InchwormHPX on top of this PR.
If I am not mistaken, the estimated autocorrelation time is 0 after 1000 measurements because `DEFAULT_MIN_SAMPLES==1024`. Convergence of `tau` is indeed observed, but for much larger numbers of measurements.
```
n measurements  tau
--------------  ---
1000000         25.83
2000000         29.08
3000000         30.34
5000000         33.12
10000000        33.78
```

**Comment by mwallerb** [comment]
Added some tests (and a little cleanup).

File: alea/test/stream_serializer.cpp [hunk 4]
```diff
@@ -142,6 +144,7 @@ class twogauss_serialize_case
 
         archive >> out; // deserialize
 
+        std::cerr << alps::alea::PRINT_VERBOSE << "\nout\n" << out << "\n";
         EXPECT_EQ(in, out);
```
Comment (galexv) [inline alea/test/stream_serializer.cpp]: (Quoting https://github.com/ALPSCore/ALPSCore/pull/477#commitcomment-27816874):
> K: Just in case, why didn't you do the same for `batch_result`?
> M: Because `batch_result` does not normalize any results, i.e., the comparison works for `count() == 0`

Sorry for harping on this again :), but how do we know it?
It would be nice to have a test verifying that `in==out` is true iff `in` and `out` are equal, for usual cases and some reasonable corner cases like `count()==0` .

Comment (mwallerb) [reply alea/test/stream_serializer.cpp]: Let's make this a seperate issue though, shall we (#482)? Otherwise this PR will drag on forever ...
